### DISCUSSION TITLE
TypeError: Cannot convert undefined or null to object

### DISCUSSION THREAD

**Discussion by donemmx** [nextauthjs/next-auth#2978] [Help]
### Description 🐜

Please help solve this issue. i only added more providers and boom it started complaining of this, and I removed the provider but this bug won't just go. i have restarted the local server several times but to no avail. please help me out

![Screenshot (78)](https://user-images.githubusercontent.com/55012848/137615814-9c03edcc-653d-4937-81fe-887e19746582.png)

![Screenshot (79)](https://user-images.githubusercontent.com/55012848/137615815-a86f69ce-cfe5-4f17-bb61-9a9a9fcceb6e.png)
![Screenshot (80)](https://user-images.githubusercontent.com/55012848/137615816-76cecb2d-cb91-4af4-9071-d80d96a3393c.png)
![Screenshot (81)](https://user-images.githubusercontent.com/55012848/137615818-00b0114c-3377-4ca8-a8a4-e721b9bdf2a4.png)

### Is this a bug in your own project?

Yes

### How to reproduce ☕️

https://next-auth.js.org/configuration/pages#sign-in-page

### Environment 🖥

 System:
    OS: Windows 10 10.0.19041
    CPU: (8) x64 Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
    Memory: 3.13 GB / 15.86 GB
  Binaries:
    Node: 16.9.1 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.5 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 7.16.0 - ~\AppData\Roaming\npm\npm.CMD
  Browsers:
    Edge: Spartan (44.19041.1266.0), Chromium (94.0.992.50)
    Internet Explorer: 11.0.19041.1202
  npmPackages:
    next: 11.1.2 => 11.1.2
    next-auth: 4.0.0-beta.4 => 4.0.0-beta.4
    react: 17.0.2 => 17.0.2

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by balazsorban44** [comment]
looks like `providers` isn't properly propagated. do you have a custom `_app`? could you show that?

**Reply by donemmx** [reply]
ok

**Comment by donemmx** [comment]
![Screenshot (82)](https://user-images.githubusercontent.com/55012848/137616798-75cdef18-142a-47ab-b5cc-786c8fef9d6e.png)

**Comment by balazsorban44** [comment]
I recommend pasting in code instead of images next time. when you log out the result of `getProviders` what do you get?

**Reply by donemmx** [reply]
It just picked up again. and it's working now. I didn't do anything or change anything.

**Reply by donemmx** [reply]
what could be the problem?

**Reply by balazsorban44** [reply]
it's very hard to say without a reproduction, the images looked good, great that it works!

**Reply by deadcoder0904** [reply]
@balazsorban44 i have the same issue now (using v3)

## _app.tsx

```tsx
import { Toaster } from 'react-hot-toast'
import type { AppProps } from 'next/app'
import { Provider as NextAuthProvider } from 'next-auth/client'
import { Provider } from 'urql'
import { client } from '@/client/graphql/client'
import Layout from '@/client/components/Layout'

function CustomApp({ Component, pageProps: { session, ...pageProps } }: AppProps) {
	return (
		<NextAuthProvider session={session}>
			<Provider value={client}>
				<Layout>
					<Component {...pageProps} />
					<Toaster />
				</Layout>
			</Provider>
		</NextAuthProvider>
	)
}

export default CustomApp
```

## navbar.tsx

```tsx
import { GetServerSideProps } from 'next'
import Link from 'next/link'
import { signIn, signOut, getProviders, ClientSafeProvider } from 'next-auth/client'

import { useGetCurrentUserQuery } from '@/client/graphql/getCurrentUser.generated'

interface IHomePage {
	providers: Record<string, ClientSafeProvider>
}

const Navbar = ({ providers }: IHomePage) => {
	const [{ data }] = useGetCurrentUserQuery()
	const isAuthenticated = !!data?.currentUser

	if (!isAuthenticated) {
		return (
			<>
				{Object.values(providers).map((provider) => (
					<button key={provider.name} onClick={() => signIn(provider.id)}>
						Login with {provider.name}
					</button>
				))}
			</>
		)
	}

	return (
		<div style={{ display: `flex`, justifyContent: `space-between` }}>
			{<Link href="/api/auth/logout">Logout</Link>}
		</div>
	)
}

export const getServerSideProps: GetServerSideProps = async () => {
	const providers = await getProviders()

	return {
		props: {
			providers,
		},
	}
}

export default Navbar
```

anything i'm doing wrong?

works in my other project but i have it under `pages/` & in this case it's under `components/navbar.tsx`. i get the following error:

> TypeError: Cannot convert undefined or null to object

**Reply by balazsorban44** [reply]
you cannot use getServerSideProps outside of `pages`

**Reply by deadcoder0904** [reply]
i think `getServerSideProps` doesn't get called outside of `pages/` so i went with the client-side method but i get different error on `Object.values`:

> No overload matches this call.
  Overload 1 of 2, '(o: { [s: string]: ClientSafeProvider; } | ArrayLike<ClientSafeProvider>): ClientSafeProvider[]', gave the following error.
    Argument of type 'Record<string, ClientSafeProvider> | null' is not assignable to parameter of type '{ [s: string]: ClientSafeProvider; } | ArrayLike<ClientSafeProvider>'.
  Overload 2 of 2, '(o: {}): any[]', gave the following error.
    Argument of type 'Record<string, ClientSafeProvider> | null' is not assignable to parameter of type '{}'.
      Type 'null' is not assignable to type '{}'.ts(2769)

**Reply by deadcoder0904** [reply]
@balazsorban44 just saw your reply haha but idk how to solve this as `Object.values` gives red-squiggly lines.

<img width="748" alt="Screenshot 2021-10-17 at 4 54 09 PM" src="https://user-images.githubusercontent.com/16436270/137625389-e1a4d66f-f0c7-4565-9eee-a0c15c49130d.png">

**Reply by balazsorban44** [reply]
you cannot call async functions in React components.

**Reply by deadcoder0904** [reply]
Just solved it:

```ts
import React from 'react'
import { GetServerSideProps } from 'next'
import Link from 'next/link'
import { signIn, signOut, getProviders, ClientSafeProvider } from 'next-auth/client'

import { useGetCurrentUserQuery } from '@/client/graphql/getCurrentUser.generated'

const Navbar = () => {
	const [{ data }] = useGetCurrentUserQuery()
	const isAuthenticated = !!data?.currentUser
	const [providers, setProviders] = React.useState<Record<string, ClientSafeProvider>>({})

	React.useEffect(() => {
		async function getProvidersValue() {
			const p = await getProviders()
			setProviders(p as Record<string, ClientSafeProvider>)
		}
		getProvidersValue()
	}, [])

	if (!isAuthenticated)
		return (
			<>
				{Object.values(providers).map((provider) => (
					<button key={provider.name} onClick={() => signIn(provider.id)}>
						Login with {provider.name}
					</button>
				))}
			</>
		)

	return (
		<div style={{ display: `flex`, justifyContent: `space-between` }}>
			<Link href="/api/auth/logout">Logout</Link>
		</div>
	)
}

export default Navbar
```

**Reply by robertomaster** [reply]
I solved this issue coping my `singin.js` file into a folder called `auth` into `pages` folder. Like this:

![image](https://user-images.githubusercontent.com/45952206/197347574-931771c6-5807-4e2b-b16c-877264b4bf0d.png)

[Here ](https://github.com/ndom91/next-auth-example-sign-in-page)a example signin page github repository, where I found this solution.

I hope I helped
